nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java

```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
-        // The nullness of one argument is all that matters for the antecedent, let's negate the
-        // consequent to fix the nullness of this argument.
-        AccessPath accessPath =
-            AccessPath.getAccessPathForNodeNoMapGet(node.getArgument(argIdx), apContext);
-        if (accessPath == null) {
-          continue;
-        }
-        if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NULLABLE)) {
-          // If argIdx being null implies the return of the method being false, then the return
-          // being true implies argIdx is not null and we must mark it as such in the then update.
-          thenUpdates.set(accessPath, Nullness.NONNULL);
-        } else if (consequent.equals("true") && argAntecedentNullness.equals(Nullness.NULLABLE)) {
-          // If argIdx being null implies the return of the method being true, then the return being
-          // false implies argIdx is not null and we must mark it as such in the else update.
-          elseUpdates.set(accessPath, Nullness.NONNULL);
-        } else if (consequent.equals("fail") && argAntecedentNullness.equals(Nullness.NULLABLE)) {
-          // If argIdx being null implies the method throws an exception, then we can mark it as
-          // non-null on both non-exceptional exits from the method
-          bothUpdates.set(accessPath, Nullness.NONNULL);
+      }
+      if (!supported) {
+        // Too many arguments involved, or unsupported @Contract features. On to next clause in
+        // the
+        // contract expression
+        continue;
+      }
+      if (arg == null) {
+        // The antecedent is unconditionally true. Check for the ... -> !null case and set the
+        // return nullness accordingly
+        if (consequent.equals("!null")) {
+          return NullnessHint.FORCE_NONNULL;
         }
+        continue;
+      }
+      if (argAntecedentNullness == null) {
+        throw new IllegalStateException("argAntecedentNullness should have been set");
+      }
+      // The nullness of one argument is all that matters for the antecedent, let's negate the
+      // consequent to fix the nullness of this argument.
+      AccessPath accessPath = AccessPath.getAccessPathForNodeNoMapGet(arg, apContext);
+      if (accessPath == null) {
+        continue;
+      }
+      if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NULL)) {
+        // If arg being null implies the return of the method being false, then the return
+        // being true implies arg is not null and we must mark it as such in the then update.
+        thenUpdates.set(accessPath, Nullness.NONNULL);
+      } else if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // If arg being non-null implies the return of the method being false, then the return
+        // being true implies arg may be null and we must mark it as such in the then update.
+        thenUpdates.set(accessPath, Nullness.NULLABLE);
```

Comment:
> There are two options when visiting the statement `foo(o)`, either a) `o` is nullable, or b) `o` is non-null (`o` being `Nullness.NULL` is a case of being nullable, but also exceedingly rare in our tracking). If `o` was non-null before `foo(o)`, it will remain so after. If it was nullable, then there is no point to update it to `Nullness.NULLABLE`.
> 
> So, in which case would we need this change to `thenUpdates`?
> 
> (Generally, it makes sense to mark something that was nullable previous to a method call as now known to be non-null, but it makes no sense to mark an _existing_ AP as being nullable unless the value at that AP was somehow updated, I think...)

This is exactly the issue I've been thinking about.  I think the only time an AP can go from being non-null to nullable is if it gets mutated by the callee method.  This is impossible for locals, but possible for field accesses / method call results.  NullAway unsoundly assumes invoked methods are side-effect-free and deterministic, and I think it would be very surprising to users if `@Contract` annotations were capturing side effects and leading to error messages that wouldn't be there otherwise.  In general, I think we should aim for our `@Contract` handling to strictly remove false positives, not to introduce new true positives.